### DISCUSSION TITLE
Can nodes be added programatically?

### DISCUSSION THREAD

**Discussion by nutmix** [vakata/jstree#2708] [General]
we have a very complex data structure we pull from various APIs.  We want to traverse this structure and create nodes and sub nodes etc.  Also, if he user tries to open a particular node, we will need a callback which will inject further nodes after making various api calls and processing the data.  I see various methods to get nodes, but none to "add" them?

**Comment by vakata** [comment]
You probably need to set core.data to a callback function and make your requests / api calls on the server and return as needed. You can also use create_node(), but from what I understand you will be better suited to lazy loading the tree.
